src/manifest/pfm.rs

```diff
@@ -0,0 +1,497 @@
+    }
+
+    /// Returns the format version for the element this entry refers to.
+    ///
+    /// Along with the type of the element, this value describes how to decode
+    /// the serialized element in the PFM.
+    pub fn format_version(&self) -> u8 {
+        self.format_version
+    }
+
+    /// Returns the flash address at which this entry's element begins.
+    ///
+    /// This value is measured from the start of the PFM's manifest frame,
+    /// rather than from the start of the PFM's table of contents.
+    pub fn offset(&self) -> u16 {
+        self.offset as _
+    }
+
+    /// Returns the length, in bytes, of this entry's element.
+    pub fn len(&self) -> u16 {
+        self.len as _
+    }
+
+    /// Returns the TOC index of this entry's parent, if it has one.
+    pub fn parent_idx(&self) -> Option<usize> {
+        match self.parent_idx {
+            0xff => None,
+            x => Some(x as _),
+        }
+    }
+
+    /// Returns the TOC hash index of this entry, if it has one.
+    pub fn hash_idx(&self) -> Option<usize> {
+        match self.hash_idx {
+            0xff => None,
+            x => Some(x as _),
+        }
+    }
+}
+
+/// The table of contents of a PFM.
+///
+/// The table of contents is encoded as follows:
+/// ```text
+/// struct Toc {
+///   entry_count: u8,
+///   hash_count: u8,
+///   hash_type: u8,
+///   entries: [TocEntry; self.entry_count],
+///   hashes: [Digest; self.hash_count],
+///   table_hash: Digest,
+/// }
+/// ```
+///
+/// The layout of the `TocEntry` type is described in [`TocEntry`]. `Digest` is
+/// a hash specified by `hash_type`. Currently, Manticore does not support hash
+/// types other than SHA-256. See [`HashType`] for more information.
+///
+/// The `entries` represent the actual entries to the table of contents; each
+/// entry refers to an *element* in the body of the PFM, describing where it is
+/// and how to decode it.
+///
+/// The `hashes` are digests of certain elements, which entries in the table of
+/// contents refer to. When an element is read out of flash, its hash is
+/// verified against the one present in the TOC, if its TOC entry indicates a
+/// hash. This ensures that the element retains protection of the overall
+/// manifest signature, even if it is parsed long after the TOC has been read
+/// out of flash.
+///
+/// It is not possible to construct a `Toc` directly; it must be parsed out of
+/// a valid [`Pfm`].
+///
+/// [`TocEntry`]: struct.TocEntry.html
+/// [`HashType`]: enum.HashType.html
+/// [`Pfm`]: struct.Pfm.html
+pub struct Toc<'toc> {
+    entries: &'toc [TocEntry],
+    hashes: &'toc [sha256::Digest],
+}
+
+impl<'toc> Toc<'toc> {
+    /// Checks that all invariants of this `Toc` type hold:
+    /// - Every pointer to a hash is in-bounds.
+    /// - Every pointer to a parent is in-bounds.
+    ///
+    /// Returns true if the invariants are upheld.
+    fn verify_invariants(&self) -> bool {
```

Comment:
Is there a requirement that the PFM be acyclic?

Reply:
Not currently! Probably something to bring to Microsoft later on.